### ISSUE TITLE
Make swarm dep optional

### ISSUE THREAD

**Issue by tsutsu** [quantum-elixir/quantum-core#402]
tl;dr: When `quantum` is used with a scheduler with `global: false`, no code from `swarm` is called. Yet, `swarm` is still pulled in as a dep.

Personally, I don't use `swarm` for anything, and am using `quantum` in a single-node context. I don't want swarm's added startup delay, or its messing with the `net_kernel`. I would like to avoid including it in my release, but `quantum` is pulling it in... even though it then doesn't use it at all (in a `global: false` scheduler configuration.)

Currently I have just created a stubbed `swarm` at `vendor/swarm` and added a `{:swarm, path: "vendor/swarm", override: true}` to my project mixfile to make `swarm` go away.

One way to fix this is to make the dep on `swarm` were optional. I can see the argument against this, though—maybe one of my own deps will depend on creating a `global: true` scheduler without my knowledge, and then it'll mysteriously break in builds that don't include `swarm`. This could be partially ameliorated with extra logging for when `quantum` tries to start up in a `global: true` configuration and fails to find `swarm`. (Sort of like when Ecto tries to start up with an adapter like `Ecto.Adapters.Postgres` and doesn't find `postgrex`.)

Another option is to fix this on the `swarm` side: add a "stub resolver" strategy to `swarm` where it doesn't start up its supervision tree, and instead just acts as a transparent proxy for the regular stuff—`Swarm.Registry` passing through to `:global`, `Swarm.members` returning `[node()]`, etc. Then other projects that rely on `swarm` could also be similarly "fixed." This would also simplify all of the various `global: true` vs. `global: false` code-paths in `quantum` (that likely also exist in other projects that optionally rely on `swarm`.)

A third option (if option 2 seems like the better idea, but @bitwalker doesn't like the sound of having that additional code in `swarm`) is to create a library wrapping `swarm` that does an `included_applications: [:swarm]`, which either boots up `swarm`'s supervision tree and passes through to it, or doesn't start up anything, and instead acts as the "stub resolver" strategy described above. The benefit here is that the API of such a library could be more closely tuned to the use-case of `quantum`, such that all code-paths branching on the value of the `global` config param could likely be unified.

Thoughts?

**Comment by LostKobrakai** [comment]
I'd vote for option 1. Imho a scheduler should either be created by an application I control and where I add jobs or the dependency works totally independant and therefore must ensure it does require swarm or list it as optional and document the need for swarm just like quantum itself.

**Comment by maennchen** [comment]
I've added a partial fix to this issue with #410.

After this PR, `swarm` will only be started in global mode, is still a required dependency though.

I'm in the process of rethinking the whole situation with `swarm` and global mode which should also tackle this issue fully.

**Comment by maennchen** [comment]
Replaced by #411
